### ISSUE TITLE
Support request for user defined iptables or iptables off/false, as in Docker Engine

### ISSUE THREAD

**Issue by b1shan** [containers/libpod#4219]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

Unlike Docker Engine, Podman does not provide support for turning iptables off or false, such as in dockerd cli or config. Moreover, Podman does not provide a user chain, such as DOCKER-USER, where users can put their own rules with the guarantee that they won't be oven run by Docker Engine.

Ref:
https://docs.docker.com/network/iptables/

**Comment by mheon** [comment]
This seems like it would need to be done inside of CNI.

**Comment by b1shan** [comment]
I think having either of the option would be a good start - option to turn iptables off or provide a predefined chain for users. The Podman security model is great, a strategic improvement over Docker Engine. This feature aligns with the security model Podman is pushing for IMO. Without this feature, several users are going to have trouble migrating over to Podman, especially the case of hybrid compute environments. The internet is full of usability issues with dockerd and iptables, before they provided an alternative.

**Comment by baude** [comment]
@b1shan willing to contribute?

**Comment by baude** [comment]
@b1shan also note, this would be for rootfull containers only?

**Comment by b1shan** [comment]
@baude yes, that is an option if this isn't a priority for anyone else at the moment. I am more from the user side, haven't seen the code much or the cni lib. If you all help with pointers where to do what, I can, why not.

**Comment by b1shan** [comment]
@baude not sure I understand what you mean by rootfull containers. This is for containers exposed on the network - on the host and from outside the host.

**Comment by baude** [comment]
the networking stack is setup differently when containers are run by root vs a regular user.

**Comment by rhatdan** [comment]
Right, rootless containers do not do anything with iptables.

**Comment by mheon** [comment]
We've had more requests for this before (specifically the USER chain)

**Comment by b1shan** [comment]
Spent a few mins looking at the CNI plugins repo and it seems setting iptables off should be rather straight forward, primarily in bridge and p2p plugins. Supporting USER chain, such as CNI-USER will be a bit of work though.

**Comment by b1shan** [comment]
A user chain is certainly a cleaner approach though, especially for dynamic workloads.

**Comment by rhatdan** [comment]
@b1shan @baude Still working on this one?

**Comment by rhatdan** [comment]
No one seems to have touched this in a couple of months.  Is this still something we should have?

**Comment by mheon** [comment]
Yes, we definitely want to support this. Unfortunately the work is entirely on the CNI side.

**Comment by rhatdan** [comment]
@mccv1r0 Any chance you could look at this?

**Comment by mccv1r0** [comment]
> Yes, we definitely want to support this. Unfortunately the work is entirely on the CNI side.

Podman would need to have an on/off switch and tell CNI about it's value.  Podman would also need to supply any rules to a USER chain.   IIRC, the plan was for CNI to let runtime add what they want to firewall via `Chain CNI-ADMIN` (if not that one, another can be created.)

**Comment by rhatdan** [comment]
@mccv1r0 @mheon Any progress on this?

**Comment by rhatdan** [comment]
@baude @mheon @mccv1r0 We are still blocked by CNI?

**Comment by Luap99** [comment]
@mheon WDYT, we could implement this easily in netavark.

Add a field to containers.conf (maybe `firewall_management = true|false`) and pass this down to netavark. In netavark we could implement a new none firewall driver which implements the interface but only provides stub methods with no actual logic.

**Comment by mheon** [comment]
Isn't the ask here to continue to add firewall rules as normal, but add an additional chain that runs before any rules we generate, so that users can add their own filtering?

**Comment by mheon** [comment]
Regardless of exact details, concur that this is an easy ask for netavark

**Comment by mheon** [comment]
Ahh. Is the request to *completely* disable creation of rules? No NAT, even? That seems strange, but if people really want it, OK.

**Comment by fansari** [comment]
I am not able to start more than one container in my scenario with existing nftables.

There is always this "ERRO[0000] "netavark: code: 1, msg: iptables: Chain already exists" error.

It would be good to be able to prevent podman of touching nftables, iptables or any other firewall.

I have now described my issue here:

https://github.com/containers/netavark/issues/274

In way it is now I cannot use podman root containers anymore.

**Comment by fansari** [comment]
Today I have put this workaround into the systemd service files which start my containers:

```
ExecStartPre=/usr/bin/systemctl reload nftables                                                                                                                                                                                               
ExecStartPost=/usr/bin/systemctl reload nftables  
```

With this I am able to have running running containers after rebooting my PC.

This cleans all this extra stuff before and after the start of each container.

A better solution of course would be to be able to disable firewall changes by podman.

I never understood this philosophy anyway although I know it from docker. If docker/podman do firewall changes these changes are lost once the user reloads his firewall for any reason. 

In docker I also disable this so I have no issues after reloading my firewall.

**Comment by fansari** [comment]
As I noticed this also affects "podman build". Without stopping nftables every build fails.

I am using podman with netavark and I hope that this feature will be implemented soon.
